frontend/src/components/ViewActivities/editActivity.js

```diff
@@ -0,0 +1,157 @@
+    this.editEndDateRef = React.createRef();
+    this.editStartTimeRef = React.createRef();
+    this.editEndTimeRef = React.createRef();
+    this.editDescriptionRef = React.createRef();
+    this.editStartLocRef = React.createRef();
+    this.editEndLocRef = React.createRef();
+    this.startTz = React.createRef();
+    this.endTz = React.createRef();
+  }
+  
+  /**
+   * Edit an activity in the database upon form submission.
+   * TODO: Update times as well! This only does the text field forms (#64).
+   */
+  editActivity() {
+    let newVals = {};
+    if (this.editTitleRef.current.value !== '') {
+      newVals[DB.ACTIVITIES_TITLE] = this.editTitleRef.current.value;
+    }
+    if (this.editDescriptionRef.current.value !== '') {
+      newVals[DB.ACTIVITIES_DESCRIPTION] = this.editDescriptionRef.current.value;
+    }
+    if (this.editStartLocRef.current.value !== 'No Change'){
+      newVals[DB.ACTIVITIES_START_COUNTRY] = this.editStartLocRef.current.value;
+    }
+    if (this.editEndLocRef.current.value !== 'No Change'){
+      newVals[DB.ACTIVITIES_END_COUNTRY] = this.editEndLocRef.current.value;
+    }
+    if (Object.keys(newVals).length !== 0) {
+      writeActivity(this.props.activity.tripId, this.props.activity.id, newVals);
+    }
+  }
+
+  /** Runs when the `submit` button on the form is pressed.  */
+  finishEditActivity(event) {
+    event.preventDefault();
+    this.editActivity();
+    this.props.submitFunction();
+  }
+
+  startTimeTzUpdate = () => { this.setState({startTz : !this.state.startTz})};
+  endTimeTzUpdate = () => { this.setState({endTz : !this.state.endTz})};
+
+  /**
+   * Returns a dropdown of all the timezones.
+   * 
+   * @param st either "start" or "end" depending on whether the timezone is for the start or end timezone.
+   */
+  timezonePicker(st) {
+    return (
+      <div>
+      <Form.Control as="select"
+        ref={st === "start" ? this.startTz : this.endTz}
+        key={st === "start" ? this.state.startTz : this.state.endTz}
+      >
+        {time.timezonesForCountry(this.editStartLocRef).map((item, index) => {
+            return (<option key={index}>{item}</option>);
+          })}
+      </Form.Control>
+      </div>
+    )
+  }
+  /**
+   * Create a dropdown of all the countries.
+   * 
+   * @param ref The reference to attach to the dropdown.
+   * @param tzchange The function to run to update the time zone picker.
+   */
+  countriesDropdown(ref, tzref) {
+    return (
+      <Form.Control as="select" ref={ref} onChange={tzref}>
+        <option key="-1">No Change</option>
+        {countryList.map((item, index) => {
+          return (
+            <option key={index} eventKey={index}>{item}</option>
+          );
+        })}
+      </Form.Control>
+    );
+  }
+
+
+  render() {
+    const activity = this.props.activity;
+    const titleWidth = 3;
+    const countryWidth = 6;
+    const dateWidth = 4;
+    const timeWidth = 2;
+    const tzPickerWidth = 3;
```

Comment:
Nit: Maybe capitalize to reflect that they are deeply immutable? (unless you're planning some fancy changeable width for the future or something)

Reply:
ooh yeah I completely forgot about that